File: mobile/with-flutter/lib/features/wallets/signing/sui_signing_examples.dart
import 'dart:convert';

import 'package:para/para.dart';

import 'signing_example.dart';

abstract final class SuiSigningExamples {
  static const definitions = <SigningExampleDefinition>[
    SigningExampleDefinition(
      id: 'sui-message-plain',
      title: 'Plain personal message',
      description: 'Signs UTF-8 text with the Sui personal-message intent.',
      operation: SigningOperation.message,
    ),
    SigningExampleDefinition(
      id: 'sui-message-personal',
      title: 'Personal message bytes',
      description:
          'Signs caller-provided base64 bytes with the Sui personal-message intent.',
      operation: SigningOperation.message,
    ),
    SigningExampleDefinition(
      id: 'sui-transaction-bcs',
      title: 'Serialized BCS transaction',
      description:
          'Signs canonical base64 BCS transaction bytes with the Sui transaction intent.',
      operation: SigningOperation.transaction,
    ),
  ];

  static const _sampleTransaction =
      'AAABAAgBAAAAAAAAAAECAAEBAAAREREREREREREREREREREREREREREREREREREREREREQEiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIgEAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEREREREREREREREREREREREREREREREREREREREREREBAAAAAAAAAEBCDwAAAAAAAA==';

  static List<SigningExampleAction> build({
    required SigningExampleClient client,
    required Wallet wallet,
  }) {
    final walletId = _require(wallet.id, 'wallet ID');

    return [
      SigningExampleAction(
        definition: definitions[0],
        sign: (_) => client.signMessage(
          walletId: walletId,
          message: 'Sign in to the Para Sui example',
          chainType: BridgeChainType.sui,
        ),
      ),
      SigningExampleAction(
        definition: definitions[1],
        sign: (_) => client.signMessage(
          walletId: walletId,
          message: SuiPersonalMessage(
              base64Encode(utf8.encode('Sui personal-message bytes'))),
          chainType: BridgeChainType.sui,
        ),
      ),
      SigningExampleAction(
        definition: definitions[2],
        requiresInput: true,
        initialInput: _sampleTransaction,
        sign: (payload) => client.signTransaction(
          walletId: walletId,
          transaction: SerializedTransaction.sui(payload!),
          chainType: BridgeChainType.sui,
        ),
      ),
    ];
  }

  static String _require(String? value, String label) {
    if (value == null || value.isEmpty) {
      throw StateError('The wallet has no $label.');
    }
    return value;
  }
}

File: mobile/with-swift/example/Signing/SuiSigningExamples.swift
import Foundation
import ParaSwift

enum SuiSigningExamples {
    static let definitions = [
        SigningExampleDefinition(
            id: "sui-message-plain",
            title: "Plain personal message",
            description: "Signs UTF-8 text with the Sui personal-message intent.",
            operation: .message
        ),
        SigningExampleDefinition(
            id: "sui-message-personal",
            title: "Personal message bytes",
            description: "Signs caller-provided base64 bytes with the Sui personal-message intent.",
            operation: .message
        ),
        SigningExampleDefinition(
            id: "sui-transaction-bcs",
            title: "Serialized BCS transaction",
            description: "Signs canonical base64 BCS transaction bytes with the Sui transaction intent.",
            operation: .transaction
        ),
    ]

    private static let sampleTransaction =
        "AAABAAgBAAAAAAAAAAECAAEBAAAREREREREREREREREREREREREREREREREREREREREREQEiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIgEAAAAAAAAAIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAEREREREREREREREREREREREREREREREREREREREREREBAAAAAAAAAEBCDwAAAAAAAA=="

    static func build(manager: ParaManager, wallet: Wallet) -> [SigningExampleAction] {
        [
            SigningExampleAction(definition: definitions[0]) { _ in
                try signingResultDescription(await manager.signMessage(
                    walletId: wallet.id,
                    message: "Sign in to the Para Sui example",
                    chainType: .sui
                ))
            },
            SigningExampleAction(definition: definitions[1]) { _ in
                let message = SuiPersonalMessage(
                    data: Data("Sui personal-message bytes".utf8).base64EncodedString()
                )
                return try signingResultDescription(await manager.signMessage(
                    walletId: wallet.id,
                    message: message,
                    chainType: .sui
                ))
            },
            SigningExampleAction(
                definition: definitions[2],
                requiresInput: true,
                initialInput: sampleTransaction
            ) { payload in
                let serializedPayload = try requireInput(
                    payload,
                    label: "Canonical base64 payload",
                    for: definitions[2].title
                )
                return try signingResultDescription(await manager.signTransaction(
                    walletId: wallet.id,
                    transaction: SerializedTransaction.sui(serializedPayload),
                    chainType: .sui
                ))
            },
        ]
    }
}
